docs/ProfilingDevelopment.md

```diff
@@ -140,3 +142,153 @@ available, from the tracer.
+(And threads not in the main Ractor don't get this marker so this is one way we filter them out.)
+
+Could we have stored a pointer to the thread context directly on the thread? Potentially, yes, but we'd need to be
+extremely careful when accessing thread contexts and when cleaning them up. (Maybe we'll evolve in this direction in
+the future?)
+
+With the storage problem solved, here's what happens: the first part is that `on_gvl_waiting` records a timestamp for
+when waiting started in the thread in `per_thread_gvl_waiting_timestamp`.
+
+Then, `on_gvl_running` checks the duration of the waiting (e.g. time between waiting started and current time). This
+is a mechanism for reducing overhead: we'll produce at least two samples for every "Waiting for GVL" event that
+we choose to sample, so we need to be careful not to allow situations with a lot of threads waiting for very brief
+periods to induce too much overhead on the application.
+
+If we decide to sample (duration >= some minimal threshold duration for sampling), we can't sample yet!
+As documented above, `on_gvl_running` is called in response to VM `RUBY_INTERNAL_THREAD_EVENT_RESUMED` events that
+happen right after the thread acquired the GVL but there's still some book-keeping to do. Thus, we don't sample
+from this method, but use the `bool` return to signal the caller when a sample should be taken.
+
+Then, a sample is taken once the caller (`CpuAndWallTimeWorker`) calls into `sample_after_gvl_running`.
+This design is similar to GC profiling, where we also can't sample during GC, and thus we trigger a sample to happen immediately after.
+
+### Representing the Waiting for GVL in `ThreadContext`
+
+As far as sampling goes, we represent a Waiting for GVL as a thread state, similar to other thread states we emit.
+This state is special, as it "overrides" other states, e.g. if a thread was sleeping, but wants to wake up, even
+though the thread is still inside `sleep`, it will have the "Waiting for GVL" state.
+
+Waiting for GVL does not affect the regular flamegraph, only the timeline visualization, as it's a thread state, and
+currently we do not represent thread states in any way on the regular flamegraph.
+
+The timestamp of the beginning of waiting for GVL gets used to create a sample that represents the "Waiting for GVL"
+period. Because there's some unaccounted for cpu and wall-time between the previous sample and the start of a waiting
+period, we also create a sample to account for this.
+
+There's some (lovely?) ASCII art in `handle_gvl_waiting` to explain how we create these two samples.
+
+Once a thread is in the "Waiting for GVL" state, then all regular cpu/wall-time samples triggered by the `CpuAndWallTimeWorker`
+will continue to mark the thread as being in this state, until `on_gvl_running` + `sample_after_gvl_running` happen and
+clear the `per_thread_gvl_waiting_timestamp`, which will make samples revert back to the regular behavior.
```

Comment:
From my read of the VM sources, it should be immediate.

Looking at the VM sources `RUBY_INTERNAL_THREAD_EVENT_RESUMED` gets called inside [`thread_sched_wait_running_turn`](https://github.com/ruby/ruby/blob/aa2662d5906cb3d7b55b42539b39976f5a32856a/thread_pthread.c#L916). Following a bit along the chain, that method gets called from `thread_sched_to_running_common` <- `thread_sched_to_running`, etc.

At some point, we end up in `BLOCKING_REGION`. This method gets called by `rb_nogvl`, `rb_thread_io_blocking_call`, `do_select`, `rb_thread_wait_for_sngle_fd`, etc: all methods where the VM needs to block the thread. All of these methods where `BLOCKING_REGION` gets used have somewhere shortly after `BLOCKING_REGION` ends some kind of check for interruptions. Another way to get to `thread_sched_to_running` is `thread_blocking_end`, that gets used by `native_sleep` which also checks for interruptions.

Since there's a bunch of paths, I'm not 100% sure there's not some weird path where interruptions are not checked, but it seems like in almost all situations the VM checks for interruptions before going back to process whatever it was waiting for.

---

Additionally, it's not possible for 

>  a lot of cpu/wall samples get misattributed as waiting on gvl?

because once `on_gvl_running` gets called, and the timestamp is flipped to negative, the next sample (even if it's not from `sample_after_gvl_running` but a regular sample) will clear the `on_gvl_running`.

So even if we hit a weird corner case where `sample_after_gvl_running` does not run in a timely manner, the next regular sample will anyway clear the state.

Thus, this makes me think we're mostly covered ;)